File: app/src/main/java/com/vladislav/workoutassistant/ui/fragments/TimePickerFragment.java
package com.vladislav.workoutassistant.ui.fragments;

import android.app.Activity;
import android.app.Dialog;
import android.app.TimePickerDialog;
import android.content.Intent;
import android.os.Bundle;
import android.text.format.DateFormat;
import android.widget.TimePicker;

import java.sql.Time;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;

public class TimePickerFragment extends DialogFragment {

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Date time = (Date) getArguments().getSerializable(ARG_TIME);
        if (time == null) {
            time = new Date();
        }

        Calendar calendar = Calendar.getInstance();
        calendar.setTime(time);

        int hour = calendar.get(Calendar.HOUR_OF_DAY);
        int minute = calendar.get(Calendar.MINUTE);

        return new TimePickerDialog(getActivity(),
                new TimePickerDialog.OnTimeSetListener() {
                    @Override
                    public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
                        Calendar calendar = new GregorianCalendar(0,0 ,0, hourOfDay, minute);
                        sendResult(Activity.RESULT_OK, new Time(calendar.getTime().getTime()));
                    }
                },
                hour, minute, DateFormat.is24HourFormat(getActivity()));
    }

    private void sendResult(int resultCode, Date date) {
        Fragment targetFragment = getTargetFragment();
        if (targetFragment != null) {
            Intent intent = new Intent();
            intent.putExtra(EXTRA_TIME, date);
            targetFragment.onActivityResult(getTargetRequestCode(), resultCode, intent);
        }
    }

    public static TimePickerFragment newInstance(Date date) {
        Bundle args = new Bundle();
        args.putSerializable(ARG_TIME, date);

        TimePickerFragment fragment = new TimePickerFragment();
        fragment.setArguments(args);

        return fragment;
    }

    private static final String ARG_TIME = "arg_time";
    public static final String EXTRA_TIME = "extra_time";
}

File: core-components/src/main/java/com/vladislavmyasnikov/core_components/view/TimePickerFragment.kt
package com.vladislavmyasnikov.core_components.view

import android.app.Activity
import android.app.Dialog
import android.app.TimePickerDialog
import android.content.Intent
import android.os.Bundle
import android.text.format.DateFormat
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import java.sql.Time
import java.util.*

class TimePickerFragment : DialogFragment() {

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val initialTime = arguments!!.getSerializable(TIME_ARG) as Date
        val calendar = Calendar.getInstance().apply { time = initialTime }

        return TimePickerDialog(
                activity,
                TimePickerDialog.OnTimeSetListener { _, hourOfDay, minute ->
                    val newDate = GregorianCalendar(0, 0, 0, hourOfDay, minute)
                    sendResult(Activity.RESULT_OK, Time(newDate.time.time))
                },
                calendar.get(Calendar.HOUR_OF_DAY),
                calendar.get(Calendar.MINUTE),
                DateFormat.is24HourFormat(activity)
        )
    }

    private fun sendResult(resultCode: Int, date: Date) {
        targetFragment?.let {
            val intent = Intent().apply {
                putExtra(TIME_EXTRA, date)
            }
            it.onActivityResult(targetRequestCode, resultCode, intent)
        }
    }

    companion object {
        private const val TIME_EXTRA = "time_extra"
        private const val TIME_ARG = "time_arg"

        fun newInstance(time: Date): TimePickerFragment {
            return TimePickerFragment().apply {
                arguments = Bundle().apply {
                    putSerializable(TIME_ARG, time)
                }
            }
        }

        fun extractTime(data: Intent): Time {
            return data.getSerializableExtra(TIME_EXTRA) as Time
        }
    }
}
